python-package/lightgbm/callback.py

```diff
@@ -283,13 +293,15 @@ def _init(env: CallbackEnv) -> None:
     def _final_iteration_check(env: CallbackEnv, eval_name_splitted: List[str], i: int) -> None:
         nonlocal best_iter
         nonlocal best_score_list
+        nonlocal inited
         if env.iteration == env.end_iteration - 1:
             if verbose:
                 best_score_str = '\t'.join([_format_eval_result(x) for x in best_score_list[i]])
                 _log_info('Did not meet early stopping. '
                           f'Best iteration is:\n[{best_iter[i] + 1}]\t{best_score_str}')
                 if first_metric_only:
                     _log_info(f"Evaluated only: {eval_name_splitted[-1]}")
+            inited = False
```

Comment:
It seems that `inited` is reset to `False` only when early stopping is not triggered? Because
1. `inited` is reset to `False` only when `env.iteration == env.end_iteration - 1`
2. `_final_iteration_check` is not called when early stopping is triggered according to
https://github.com/microsoft/LightGBM/blob/00f87c52ec24b1f8358c230646685459e31b60b8/python-package/lightgbm/callback.py#L306-L327

Reply:
@shiyu1994 Thanks for your review!

> It seems that `inited` is reset to `False` only when early stopping is not triggered?

I believe `inited` is reset to `False` when early stopping is triggered here
https://github.com/microsoft/LightGBM/blob/dff622c26aceb5bc602a6279d4a45d7e99d389e5/python-package/lightgbm/callback.py#L339-L340
Triggering early stopping means to raise `EarlyStopException` and `inited` is reset to `False` right before that line.

Also, if it is not reset, the test `test_grid_search()` will fail because we use constant custom evaluation metric there which force early stopping to happen at `stopping_rounds` iteration because there is no improvement after the first iteration.